pkg/controller/deployment/deployment_controller.go

```diff
@@ -453,3 +472,55 @@ func (dc *DeploymentController) clearDeploymentOverlap(deployment *extensions.De
 	delete(deployment.Annotations, util.OverlapAnnotation)
 	return dc.client.Extensions().Deployments(deployment.Namespace).UpdateStatus(deployment)
 }
+
+// progressWorker runs a worker thread that pops items out of a secondary queue, checks if they
+// have failed progressing and if so it adds them back to the main queue.
+func (dc *DeploymentController) progressWorker() {
+	work := func() bool {
+		key, quit := dc.progressQueue.Get()
+		if quit {
+			return true
+		}
+		defer dc.progressQueue.Done(key)
+
+		needsResync, err := dc.checkForProgress(key.(string))
+		if err == nil && needsResync {
+			dc.queue.AddRateLimited(key)
+		}
+		dc.progressQueue.Forget(key)
+		return false
+	}
+
+	for {
```

Comment:
should be `for !work() {` but I don't feel strong for these changes. We should probably have a style guide for controllers.